Small k8s fixes and optimizations

aanm opened the pull request:
Read per commit basis

aanm committed:
k8s/watchers: set BlockOwnerDeletion to true in CEP

This field should be set to true so that k8s will block the deletion of
Pod until the CEP is also removed by k8s.

Fixes: 08dc8ca31968 ("pkg/endpoint: set Pod as the endpoint owner of a CiliumEndpoint")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
install/kubernetes: set the brief header when request for /healthz

Similar to what done with cilium status --brief, we should perform
the HTTP request by sending the same headers set with
`cilium status --brief` to Cilium server.

Fixes: d613dea5d524 ("install/kubernetes: use HTTP for agent {liveness,readiness}Probe")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint: optimize GetCiliumEndpointStatus

To get a GetCiliumEndpointStatus the function does not need to get an
endpoint model has it is expensive in terms of CPU and memory to do it
so. Thus, GetCiliumEndpointStatus is now optimized to only generate the
api Models for the fields that it actually needs.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
operator/watchers: also store ResourceVersion

If the resource version is not stored, the watcher won't be able to keep
up the last resource version it has seen for this object type so it will
assume that any event received from kube-apiserver is an event that
should be processed.

Fixes: 3a1bde594b61 ("types/slim: add slim packages for k8s structures")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s/types: do not store full ObjectMeta for CEPs

It's not required to store the full ObjectMeta of CEPs. This will help
to reduce the memory consumption of CEPs since a none of the
ObjectMeta fields are used when processing a CEP k8s event.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
maps/policymap: only dump keys of policy map

When the endpoint is synchronizing the desired policy with the enforced
policy it only uses the policy map keys. This commit introduces a new
function that will only dump the policy map keys, not the values. This
will help to decrease the memory consumption from 32 bytes to 8 bytes
per Policy entry.

Signed-off-by: André Martins <andre@cilium.io>

tklauser commented on install/kubernetes/cilium/charts/agent/templates/daemonset.yaml:
```diff
@@ -60,6 +60,9 @@ spec:
             path: /healthz
             port: {{ .Values.global.agent.healthPort }}
             scheme: HTTP
+            httpHeaders:
```
I don't think this is strictly necessary right now, `:9876/healthz` is served by a separate minimal HTTP server which will always only return HTTP 200 or 500. But I guess it's good for forward compatibility, once we might extend this status check API.

aanm replied:
@tklauser although it is served by a minimal HTTP server there's a significant difference in the CPU consumption and memory usage by the server when setting `brief: true`

aanm committed:
agent/templates: add K8S_SERVICE_{HOST,PORT} in init container

Might be useful to pass the K8S_SERVICE_HOST and K8S_SERVICE_PORT to the
Cilium's init container.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
config/templates: allow to enable health checks with helm options

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
config/templates: configure ipam mode if neither gke, azure nor eni are enabled

This avoids a duplicated entry for the "ipam" field in case the user
sets the ipam value manually and picks a GKE mode.

Fixes: 787c2763c06f ("gke: Enable native-routing mode on GKE by default")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
config/templates: disable cnp-status update for new installations

CNP status updates don't provide any useful information and prevents
Cilium from scaling for large clusters as each CNP node update will
trigger a k8s event to all remaining nodes.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -325,7 +325,6 @@ func (e *Endpoint) GetCiliumEndpointStatus(conf EndpointStatusConfiguration) *ci
 		ExternalIdentifiers: modelStatus.ExternalIdentifiers,
 		Identity:            getEndpointIdentity(modelStatus),
 		Networking:          getEndpointNetworking(modelStatus),
-		State:               compressEndpointState(modelStatus.State),
```
This was not the intent. The intent was to publish a state that doesn't change consistently. Why are you making this change? Did you observe many updates? Or are you observing another PATCH required after the initial CREAT?

aanm replied:
Reverted

tgraf commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -464,7 +464,7 @@ data:
   # A space separated list of iptables chains to disable when installing feeder rules.
   disable-iptables-feeder-rules: {{ .Values.global.disableIptablesFeederRules | join " " | quote }}
-{{ if .Values.global.ipam }}
+{{ if and .Values.global.ipam (not (or .Values.global.gke.enabled ( or .Values.global.eni .Values.global.azure.enabled ))) }}
```
Why this change? Doesn't this lead to a secondary `ipam:` line? Both Azure and ENI have a separate section which also sets `ipam:`. I'm fine with this but then you have to remove that other one.

aanm replied:
@tgraf I think you missed the "not"? This commit is fixing the "Both Azure and ENI have a separate section which also sets ipam:" since that's the current behavior.

tgraf commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -509,3 +509,6 @@ global:
     enabled: true
+
+  cnpStatusUpdates:
+    enabled: false
```
This needs a note in the upgrade guide

aanm replied:
Please take a look at the doc changes.

tgraf commented on Documentation/install/upgrade.rst:
```diff
@@ -340,6 +340,13 @@ IMPORTANT: Changes required before upgrading to 1.8.0
   .. note:: The ENI IPAM mode automatically derives the native routing CIDR so
             no action is required.
 
+* To make the usage of Cilium Network Policies (CNP) scalable for clusters with
```
This doesn't say that we changed this.

I would say something like this:
> The CiliumNetworkPolicy Status includes information which allows to derive when all nodes in a cluster are enforcing a particular CiliumNetworkPolicy. For large clusters running CRD mode, this visibility is costly as it requires all nodes to participate. In order to ensure scalability, CiliumNetworkPolicy status visibility has been disabled for all new deployments. If you want to enable it, set the ConfigMap option ``disable-cnp-status-updates`` to ``false`` by using Helm ``--set global.cnpStatusUpdates.enabled=true`` or by editing the ConfigMap directly.

joestringer commented on pkg/maps/policymap/policymap_privileged_test.go:
```diff
@@ -101,6 +101,32 @@ func (pm *PolicyMapTestSuite) TestPolicyMapDumpToSlice(c *C) {
+
+	dump, err := testMap.DumpKeysToSlice()
+	c.Assert(err, IsNil)
+	c.Assert(len(dump), Equals, 1)
```
```
[2020-05-20T21:36:16.370Z]   	 === RUN   Test
[2020-05-20T21:36:16.370Z]   	 START: policymap_privileged_test.go:130: PolicyMapTestSuite.TestDeleteNonexistentKey
[2020-05-20T21:36:16.370Z]   	 PASS: policymap_privileged_test.go:130: PolicyMapTestSuite.TestDeleteNonexistentKey	0.000s
[2020-05-20T21:36:16.370Z]   	 
[2020-05-20T21:36:16.370Z]   	 START: policymap_privileged_test.go:104: PolicyMapTestSuite.TestPolicyMapDumpKeysToSlice
[2020-05-20T21:36:16.370Z]   	 PASS: policymap_privileged_test.go:104: PolicyMapTestSuite.TestPolicyMapDumpKeysToSlice	0.000s
[2020-05-20T21:36:16.370Z]   	 
[2020-05-20T21:36:16.370Z]   	 START: policymap_privileged_test.go:78: PolicyMapTestSuite.TestPolicyMapDumpToSlice
[2020-05-20T21:36:16.370Z]   	 policymap_privileged_test.go:87:
[2020-05-20T21:36:16.370Z]   	     c.Assert(len(dump), Equals, 1)
[2020-05-20T21:36:16.370Z]   	 ... obtained int = 2
[2020-05-20T21:36:16.370Z]   	 ... expected int = 1
[2020-05-20T21:36:16.370Z]   	 
[2020-05-20T21:36:16.370Z]   	 FAIL: policymap_privileged_test.go:78: PolicyMapTestSuite.TestPolicyMapDumpToSlice
[2020-05-20T21:36:16.370Z]   	 
[2020-05-20T21:36:16.370Z]   	 OOPS: 2 passed, 1 FAILED
[2020-05-20T21:36:16.370Z]   	 --- FAIL: Test (0.00s)
[2020-05-20T21:36:16.370Z]   	 FAIL
[2020-05-20T21:36:16.370Z]   	 coverage: 5.2% of statements in ./...
[2020-05-20T21:36:16.370Z]   	 FAIL	github.com/cilium/cilium/pkg/maps/policymap	0.077s
[2020-05-20T21:36:16.370Z]   	 FAIL
[2020-05-20T21:36:16.370Z]   	 Makefile:142: recipe for target 'tests-privileged' failed
[2020-05-20T21:36:16.370Z]   	 make: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium'
[2020-05-20T21:36:16.370Z]   	 
[2020-05-20T21:36:16.370Z]   Stderr:
[2020-05-20T21:36:16.370Z]    	 make: *** [tests-privileged] Error 1
```

jrajahalme reviewed (approved):
Nice for a change that all the commits were independent of each other :-)
